src/debug_utils-inl.h

```diff
@@ -27,13 +27,38 @@ struct ToStringHelper {
   }
   static std::string Convert(const std::string& value) { return value; }
   static std::string Convert(bool value) { return value ? "true" : "false"; }
+  template <unsigned BASE_BITS,
+            typename T,
+            typename std::enable_if<std::is_integral<T>::value, int>::type = 0>
+  static std::string BaseConvert(T value, bool upper) {
+    std::string ret;
+    const char* digits = upper ? "0123456789ABCDEF" : "0123456789abcdef";
+    do {
+      unsigned digit = value & ((1 << BASE_BITS) - 1);
+      ret.push_back(
+          (BASE_BITS < 4 ? static_cast<char>('0' + digit) : digits[digit]));
+    } while ((value >>= BASE_BITS) != 0);
+    std::reverse(ret.begin(), ret.end());
+    return ret;
+  }
+  template <unsigned BASE_BITS,
+            typename T,
+            typename std::enable_if<!std::is_integral<T>::value, int>::type = 0>
+  static std::string BaseConvert(T value, bool upper) {
+    return Convert(std::forward<T>(value));
+  }
 };
 
 template <typename T>
 std::string ToString(const T& value) {
   return ToStringHelper::Convert(value);
 }
 
+template <unsigned BASE_BITS, typename T>
+std::string ToBaseString(T&& value, bool upper = false) {
+  return ToStringHelper::BaseConvert<BASE_BITS>(std::forward<T>(value), upper);
+}
```

Comment:
I don’t think you need to do any templating for `T`, any value that’s passed to a function like this will be convertible to `unsigned long long` anyway.

Reply:
Of course ok, but we had code more template for other types if we did this.

btw, I inspired from this https://github.com/fmtlib/fmt/blob/5d32ccfc3130fdd122c344b37087ac7fd8c2b62e/include/fmt/format.h#L901-L913